### PR TITLE
test: social sign on integration test

### PR DESCRIPTION
**Pull request by danielgent**
I think this has enough changes for now!

I've got some ideas for more tests on future PRs

### CODE REVIEW

File: src/authentication-flows/social.ts [hunk 1]
```diff
@@ -78,7 +78,10 @@ export async function socialAuthCallback({
   code,
 }: socialAuthCallbackParams) {
   const { env } = ctx;
-  const client = await getClient(env, state.authParams.client_id);
+  const client = await ctx.env.data.clients.get(state.authParams.client_id);
```
Comment (danielgent) [inline src/authentication-flows/social.ts:81]: need to use data adapters so our integration tests work!

File: integration-test/test-server.ts [hunk 2]
```diff
@@ -3,6 +3,63 @@ import { app } from "../src/app";
...
+  AuthorizationResponseMode,
+  AuthorizationResponseType,
+} from "../src/types";
+import {
+  IOAuth2ClientFactory,
+  OAuthProviderParams,
+  IOAuth2Client,
+} from "../src/services/oauth2-client";
+
+class MockOAuth2Client implements IOAuth2Client {
```
Comment (danielgent) [inline integration-test/test-server.ts:16]: :smile: :clown_face: 

This is the magic here!

Given we're passing this into the context I figure the idea was to then mock this?

It makes the integration tests way easier, at the expense of doing `expect(true).toBe(true)`

I figure that the OAuth2Client isn't the most complex class, and if we use it then we'll have to intercept `fetch`, or run `no-auth` on its own server and port

_instead_ using this we can test
* SSO `id_tokens` are decoded and used
* the PKCE call is kind of done in `/callback`
* new users are created with correct SSO ids
* existing users are used
* account linking happens!

These tests will catch some bugs we've had and be able to test the last of our functionality

Comment (danielgent) [reply integration-test/test-server.ts:16]: I'd be very interested in hosting NoAuth somewhere _BUT_ then using that our login2 playwright tests - maybe have a new SSO button with a log :no_entry_sign: :stuck_out_tongue:

File: integration-test/test-server.ts [hunk 3]
```diff
@@ -1,8 +1,52 @@
...
+  }
+
+  getAuthorizationUrl(state: string): Promise<string> {
+    throw new Error("getAuthorizationUrl method not implemented.");
+  }
+  exchangeCodeForTokenResponse(code: string): Promise<any> {
+    return new Promise((resolve, reject) => {
+      resolve({
+        access_token: "accessToken",
+        // copy-pasted from chat GPT into jwt.io - should create programmatically here?
```
Comment (danielgent) [inline integration-test/test-server.ts:31]: I'll do this next and create the `id_token` from the JSON so we can see what is happening

File: integration-test/flows/social.spec.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        isSocial: false,
+      },
+    ]);
+    expect(newSocialUser.user_id).toBe("demo-social-provider|1234567890");
+
+    // NEXT TESTS
+    // - account linking
+    // - logging in SSO with existing user
+    // - POST to /callback endpoint e.g. Apple SSO
+    // - malicious call to /callback! Are we verifying anything here?
```
Comment (danielgent) [inline integration-test/flows/social.spec.ts:215]: If this is all vaguely ok then I'll continue with these on a new PR :muscle: 

This test at least would have caught the bug we had with not allowing new social sign ups

File: src/services/datadog.ts [hunk 5]
```diff
@@ -20,7 +20,6 @@ async function log(
   const ddApiKey = ctx.env.DD_API_KEY;
 
   if (!ddApiKey?.length) {
-    console.log("No datadog api key available");
```
Comment (danielgent) [inline src/services/datadog.ts:23]: As mentioned in another PR, I'm just removing this :+1:

File: integration-test/flows/social.spec.ts [hunk 6]
```diff
@@ -0,0 +1,299 @@
+import { setup } from "../helpers/setup";
+import { start } from "../start";
+import { parseJwt } from "../../src/utils/parse-jwt";
+import { getAdminToken } from "../helpers/token";
+
+// same on each test
```
Comment (danielgent) [inline integration-test/flows/social.spec.ts:6]: We could have these different for each test *BUT*
* on each test the data adaptors are reset (and I test that!)
* then the mock OAuth2Provider would need to return different tokens...   I could do this by using multiple `clientId`s in a way that doesn't require a change in the architecture - and swap based on clientId...  or have different codes :upside_down_face:   There are a few ways to do it but they all involve logic

File: integration-test/flows/social.spec.ts [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+
+  beforeEach(async () => {
+    worker = await start();
+  });
+
+  afterEach(() => {
+    worker.stop();
+  });
+
+  it("should create correct args for social sign on from hitting /authorize with connection", async () => {
```
Comment (danielgent) [inline integration-test/flows/social.spec.ts:54]: I'm doing separate tests now

File: integration-test/flows/social.spec.ts [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+        updated_at,
+        last_login,
+        ...newSocialUserWithoutDates
+      } = newSocialUser;
+
+      expect(newSocialUserWithoutDates).toEqual(EXPECTED_NEW_USER);
+    });
+
+    // like apple
+    it("POST to /callback", async () => {
```
Comment (danielgent) [inline integration-test/flows/social.spec.ts:166]: this is new. surprisingly it worked

File: src/authentication-flows/social.ts [hunk 9]
```diff
@@ -149,10 +158,10 @@ export async function socialAuthCallback({
       name: email,
       provider: state.connection,
       connection: state.connection,
-      email_verified: false,
+      email_verified: strictEmailVerified,
       last_ip: "",
       login_count: 0,
-      is_social: false,
+      is_social: true,
```
Comment (danielgent) [inline src/authentication-flows/social.ts:164]: this was a bug that I have now fixed. of course this should be true :smile:

File: src/authentication-flows/social.ts [hunk 10]
```diff
@@ -162,13 +171,9 @@ export async function socialAuthCallback({
   ctx.set("email", email);
   ctx.set("userId", user.id);
 
-  const newUserFields: Partial<BaseUser> = {
-    profileData: JSON.stringify(profileData),
-    email,
-  };
-
+  // this checks everytime we get the id_token that the email is verified, and updates the user in the db
+  // possibly excessive and we could just set email_verified:true when creating the new social user (above)
```
Comment (danielgent) [inline src/authentication-flows/social.ts:175]: now I'm thinking I could nuke the following lines. 

Maybe this is all excessive and we just set `email_verified: true` for all social users initially and never change it

File: integration-test/flows/social.spec.ts [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+      const {
+        created_at,
+        updated_at,
+        last_login,
+        ...newSocialUserWithoutDates
+      } = newSocialUser;
+
+      expect(newSocialUserWithoutDates).toEqual(EXPECTED_NEW_USER);
+    });
+  });
```
Comment (danielgent) [inline integration-test/flows/social.spec.ts:250]: I've got some ideas for more tests but this test is long enough :muscle:

File: integration-test/flows/social.spec.ts (hunk 8 above)
Comment (markusahlstrand) [reply integration-test/flows/social.spec.ts:166]: Someone told me this should read as a sentence :)

Comment (danielgent) [reply integration-test/flows/social.spec.ts:166]: Fire that guy, I'll make the fix though :clown_face:

Comment (danielgent) [reply integration-test/flows/social.spec.ts:166]: I hope these tests made sense though! They give me a *LOT* more code confidence in our social logins, and they're much much quicker to run than playwright tests

If we do end up using SQLite then that would really be :100: :rocket:

**Pull request merged**